// excelcsv/Gruntfile.js
/*
 * excelcsv
 *
 *
 * Copyright (c) 2014 Jason Pescione
 * Licensed under the MIT license.
 */

'use strict';

module.exports = function (grunt) {
  // load all npm grunt tasks
  require('load-grunt-tasks')(grunt);

  // Project configuration.
  grunt.initConfig({

    watch: {
      all: {
        files: ['lib/**', 'test/**', 'Gruntfile.js', 'package.json', '.jshintrc'],
        tasks: ['jshint']
      }
    },

    jshint: {
      all: [
        'Gruntfile.js',
        'test/**/*.js',
        'lib/**/*.js'
      ],
      options: {
        jshintrc: '.jshintrc',
        reporter: require('jshint-stylish')
      }
    },

    // Before generating any new files, remove any previously-created files.
    clean: {
      tests: ['tmp']
    },

    mkdir: {
      all: {
        options: {
          create: ['tmp']
        }
      }
    },

    // Unit tests.
    mochacli: {
      options: {
        require: ['should'],
        bail: true
        // ui: 'bdd'
      },
      all: ['test/**/*.js']
    }

  });

  // Actually load this plugin's task(s).
  grunt.loadTasks('tasks');

  // Whenever the "test" task is run, first clean the "tmp" dir, then run this
  // plugin's task(s), then test the result.
  grunt.registerTask('test', ['clean', 'mkdir', 'mochacli']);

  // By default, lint and run all tests.
  grunt.registerTask('default', ['jshint', 'test']);

};
